src/main/java/edu/unc/lib/boxc/migration/cdm/services/SourceFilesToRemoteService.java

```diff
@@ -31,58 +38,90 @@ public class SourceFilesToRemoteService {
      * @throws IOException
      */
     public void transferFiles(Path destinationPath) throws IOException {
-        var sourceMappings = sourceFileService.loadMappings();
-        final Path destinationBasePath = destinationPath.toAbsolutePath();
-        // Get all the source paths as a thread safe queue
-        var pathsDeque = sourceMappings.getMappings().stream()
-                .map(SourceFilesInfo.SourceFileMapping::getFirstSourcePath)
-                .collect(Collectors.toCollection(ConcurrentLinkedDeque::new));
-        // For tracking if a parent directory has already been created
-        Set<String> createdParentsSet = ConcurrentHashMap.newKeySet();
-        // Create the remote destination directory
-        log.info("Creating remote destination directory {}", destinationBasePath);
-        sshClientService.executeRemoteCommand("mkdir -p " + destinationBasePath);
-        createdParentsSet.add(destinationBasePath.toString());
+        executor = Executors.newFixedThreadPool(concurrentTransfers);
 
-        var threads = new ArrayList<Thread>(concurrentTransfers);
-        // Start threads for parallel transfer of files
-        for (int i = 0; i < concurrentTransfers; i++) {
-            var thread = createTransferThread(pathsDeque, destinationBasePath, createdParentsSet);
-            thread.start();
-            threads.add(thread);
+        try {
+            var sourceMappings = sourceFileService.loadMappings();
+            final Path destinationBasePath = destinationPath.toAbsolutePath();
+            var pathsList = sourceMappings.getMappings().stream()
+                    .map(SourceFilesInfo.SourceFileMapping::getFirstSourcePath)
+                    .collect(Collectors.toList());
+            var pathsDeque = new ConcurrentLinkedDeque<>(pathsList);
+            // Create the parent path structure before we start transfers
+            var parentPaths = listMostSpecificParents(pathsList);
+            createParentPaths(parentPaths, destinationBasePath);
+
+            // Capture futures for transfer threads so we can receive errors
+            var futures = new ArrayList<Future<?>>();
+            for (int i = 0; i < concurrentTransfers; i++) {
+                futures.add(executor.submit(createTransferTask(pathsDeque, destinationBasePath)));
+            }
+            for (Future<?> future : futures) {
+                try {
+                    future.get();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
```

Comment:
would this just keep throwing exceptions as it loops through the futures?

Reply:
It would throw the first exception and then the whole thing would end